[Freddy CYR] iP

freddychenyouren2 opened the pull request:
# SeeWhyAre Bot 👀 👀 👀 

> An Enhanced app of the skeletal Duke bot

Hello there! 👋 This is my chatbot improved from the Duke bot 🤖 that only has one greeting line for the users.

There are lots of improvements done to this bot 🔥 . Below are the main features that have been added:
* Add and Track your task completion. Available task types: Todo, Deadline, Event.
* View your current list of task
* Filter and Find your desired tasks based on their name
* Delete tasks from the list
* Save your task data to your hard disk and load them back the next time you use SeeWhyAre bot.

### Project Setup
Want to contribute to my fork? I am more than happy to hear (or read ehe) your suggestions! Here is how you can contribute to my fork:
1. Clone my repo using the command `git clone https://github.com/freddychenyouren2/ip.git` 
2. Open IntelliJ IDEA IDE (Please install if you haven't done so. You may download the Free Community Version)
3. Configure Java SDK setup under `File` -> `Project Structure`. This project uses ****Java 11**** . 
6. In the `build.gradle` file from the root directory, click the green Play button besides the `run` command. Alternatively press `Ctrl` twice and enter `gradlew run` which should run the app.
7. Hooray! You have started SeeWhyAre bot! Happy exploring!! 😄 

### By the way...
**I hope that you have done the following:**
- [X] Download my release and play with it.
- [ ] Have a look at my [personal website](https://freddychenyouren2.wixsite.com/freddy-cyr).
- [ ] Let's connect!! I am interested in new and interesting ideas to contribute to the society! Share with me your contacts (Telegram, Whatsapp, or LinkedIn) so that I may reach out to you :)

# Java Usage
SeeWhyAre bot uses Java programming langauge, which is primarily an Object-oriented langauge. If you are not familiar with Java, here is an example block of code for you:

``` xml {.line-numbers}
public class SeeWhyAre {
    public static void main(String[] args) {
        System.out.println("Are you Seeing Why now?");
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

freddychenyouren2 committed:
Finished Level-0 increment

freddychenyouren2 committed:
Finished Level-1 increment with some personality modification for my chat bot

freddychenyouren2 committed:
Finished Level-2 Increment: Add, List

freddychenyouren2 committed:
Add a class Task to handle more complex logics involving a Task

freddychenyouren2 committed:
Finished Level-3 Incrementing: Mark as Done
Handled basic error inputs such as index out of bounds on ArrayList of Tasks

freddychenyouren2 committed:
Level-4-Part-1:
Created a Deadline.java file, handled Deadline task logics.
Corrected indentation of console text outputs.

freddychenyouren2 committed:
Level-4-part-1-Update: Handle Deadline Task in Deadline.java file instead of Duke.java
This is to practice calling Class Methods and writing neater code.
Consider if there are 100 different types of tasks. It is not good to implement Task handling modules all in Duke.java file.
It is way neater to implement task handling in each of the specific types of Task.

freddychenyouren2 committed:
Level-4-Part-2: Added Todo.java and handling a Todo Task.

freddychenyouren2 committed:
Updated Javadocs for Deadline.java fiile

freddychenyouren2 committed:
Level-4-Part-3: Added Event.java file and its logics

freddychenyouren2 committed:
A-TextUiTesting: Text Formatting correction
To prepare for automated testing

freddychenyouren2 committed:
A-TextUiTesting: Update input.txt and EXPECTED.TEXT\
To match predetermined results for outputs

freddychenyouren2 committed:
A-TextUiTesting: Use dos2unix input.txt
This is because ACTUAL.TXT and EXPECTED.TXT look the same to naked eye
The issue might lie in line-ending differences hence troubleshooting with dos2unix solved the issue.
OS: Windows 11

freddychenyouren2 committed:
Level-5: Add two classes to handle two types of exceptions
Handled basic erroneous inputs of empty todo description and Invalid command.
Might be exploring Enumeration to handle Invalid command

freddychenyouren2 committed:
Level-5-Update-1: Updated Duke.java and Todo.java
To throw the two new exceptions appropriately: EmptyDescriptionException and InvalidCommandException

freddychenyouren2 committed:
Level-6: Added deleteTask() function in Duke.java
Also added exception checkings on user inputs such as empty description, invalid index, or non-integer index.
Might explore creating Custom Exception Class in the future.

freddychenyouren2 committed:
Level-6-Update: Small corrections of coding styles

freddychenyouren2 committed:
Bug fix: "todo todo" was read as no descriptions because initial implementation wiped out ALL todo's.
Same issue for Deadline and Event objects.
Fixed with replaceFirst() instead of replace()

freddychenyouren2 committed:
A-Enum: Applied Enum on available commands and used switch-case instead of if-else.
To make code more expandable.

freddychenyouren2 committed:
Update Indentation of Switch Case

After Switch, indentation of case is provided by default by IntelliJ

Configured IntelliJ Settings to fix this indentation issue to follow course Java Coding Style Standard

freddychenyouren2 committed:
Update text messages in Exceptions

Unicode U+2369 was included in the previous version.

However, expected.txt and actual.txt files kept showing errors because the sad face icon
could not be recognized correctly. Actual.txt showed "?"

Removed Unicode U+2369 for now

freddychenyouren2 committed:
Update test file

Include more commands to test more thoroughly.

freddychenyouren2 committed:
Include Data Saving and Data Loading of tasks into hard disk

Level-7 Increment: Created a Data Directory to put duke.txt file for storing and loading user data.

However, there is one bug discovered under updateData() where Event type of task kept getting changed to Deadline type.
This will be fixed soon.

freddychenyouren2 committed:
Fix bug on Event tasks being loaded as Deadline tasks

freddychenyouren2 committed:
Implement checking of files to load data

Handled basic incorrect formats in the data file to be parsed.

freddychenyouren2 committed:
Implement Level-8 features of Date and Time in tasks

Deadline and Event tasks were initialized with String values for their timings

Used Java library on Dates and Times to initialize Deadline and Event tasks with YYYY-MM-DD inputs
Also handled cases where inputs are invalid dates or invalid formats with InvalidDateException.java

Note: LoadTasks() Function now has poorer OOP Principles used due to optional initialization of tasks read from the  hard disk.
This will be handled in the future.

freddychenyouren2 committed:
Add a comment in Todo.java file to test branching mechanism in SourceTree

No changes in any other files. This should not affect the working state of the program.

freddychenyouren2 committed:
Merge branch 'branch-level-8'

freddychenyouren2 committed:
Add a different comment in Todo.java to test branching for branch-level-7

Branching did not show up previously for branch-level-7.

This is the second attempt to practice proper branching and merging in git.

freddychenyouren2 committed:
Merge branch 'branch-level-7'

# Conflicts:
#	src/main/java/Todo.java

Solved by attempting to put the same comment in line 2

freddychenyouren2 committed:
Implemented halfway A-OOP with Ui, Storage, Parser, and TaskList Classes

Implementation of program were all under Duke.java

Extracted relevant implementations into separate classes.
Note: Main refactoring of Duke.java is unfinished. Current state is only the creation of the separate files.

freddychenyouren2 committed:
Implement A-MoreOOP by using files to separate the program.

The program was getting messy and could be separated into Parser, Ui, TaskList, and Storage files.
Duke.java file also became short and neat.

Also squashed bugs related to class method invocations especially when saving tasks or loading tasks

freddychenyouren2 committed:
Clean up code for A-MoreOOP checkpoint

Storage.java contained wildcard java.io.* import statement which violates Java Coding Standard
Duke.java contained a huge chunck of commented-out codes of previous implementations which have been removed.

freddychenyouren2 committed:
Add packages to divide classes more neatly

freddychenyouren2 committed:
Add dummy comment line to test for Merging for A-Packages

freddychenyouren2 committed:
Merge branch 'A-Packages'

freddychenyouren2 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

freddychenyouren2 committed:
Configure build.gradle file to run main method from Duke.java

freddychenyouren2 committed:
Add test files

Basic unit testing only. Future updates may see the use of stubs.

freddychenyouren2 committed:
Refactor code to adhere more closely to Java Coding Standard

freddychenyouren2 committed:
Add Header comment and JavaDoc to non-trivial methods in all Java classes.

freddychenyouren2 committed:
Implements findTask() method with a given keyword

Level-9 increment.

freddychenyouren2 committed:
no message

freddychenyouren2 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/util/Parser.java

freddychenyouren2 committed:
Merge branch 'branch-level-9'

freddychenyouren2 committed:
Refactor variable name taskList to listOfTasks in TaskList.java

This is to adhere more closely to Java Coding Standards.

freddychenyouren2 committed:
Add JavaDoc on findTask()

freddychenyouren2 committed:
Update console output of invalid commands

freddychenyouren2 committed:
Update runtest.bat to run on the correct input file for text-ui-testing

cxo05 commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,129 @@
+            try {
+                command = Command.valueOf(firstWord);
+            } catch (IllegalArgumentException e) {
+                command = Command.UNKNOWN;
```
You can immediately throw the `InvalidCommandException` here and not need an `UNKNOWN` enum

cxo05 commented on build.gradle:
```diff
@@ -0,0 +1,42 @@
+    dependsOn("distZip", "distTar")
+}
+
+run{
```
Good job on following the coding standard. This is the only styling issue I can find.

freddychenyouren2 committed:
Implement GUI partially according to JavaFX tutorial

Will be continuing to Implement GUI for the actual SeeWhyAre bot in subsequent commits.

freddychenyouren2 committed:
Update implementations of methods to return String instead of void

SeeWhyAre bot was a CLI app that used System.out.println or System.out.printf
to show messages to users.

From this commit onwawrds, it has a GUI where displaying messages requires a String.
Hence refactoring is done mainly at methods involved in parsing userInput to return a String
as the response message from SeeWhyAre bot.

freddychenyouren2 committed:
Update main storage data file

freddychenyouren2 committed:
Add negligible comment line to test for merging of other branches

freddychenyouren2 committed:
Merge branch 'branch-level-10'

freddychenyouren2 committed:
Add Java CheckStyle and refactor code to follow Java Coding Standard more

RSXIX commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,66 @@
+
+        //Need standardised photos to make better circle clips
+        dialog.setText(text);
+        //dialog.setMinHeight(Region.USE_PREF_SIZE);
```
Maybe can remove these comments if you not using

RSXIX commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,230 @@
+public class Duke {
+
+    // HORIZONTAL_LINE
+    public static String HORIZONTAL_LINE = "    ____________________________________________________________"; //60 underscores.
```
good use of variables to reduce clutter in code

RSXIX commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,230 @@
+    private Scene scene;
+
+    // Images
+    private final Image user = new Image(Objects.requireNonNull(this.getClass().getResourceAsStream("/images/ElonMusk.jpeg")));
```
Maybe can add descriptor here so we know the type that is being referred to

RSXIX commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,111 @@
+            String startTime = details[1].trim();
+            String endTime = details[2].trim();
+
+            //Check if input date is valid.
```
good use of comments to make sure the code is readable

RSXIX commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,204 @@
+     * @param task The task whose completion status is to be checked.
+     * @param completionStatus The completion status read from memory. 0 means not done, 1 means done.
+     */
+    protected void checkCompletionStatus(Task task, String completionStatus) {
```
good use of protected

RSXIX reviewed (commented):
Appreciated your use of comments but some of them felt hard to follow and I think you have alot of trial and error code left in the file which makes it feel cluttered. Coding standards seem to mostly be followed. Naming was mostly fine.

freddychenyouren2 committed:
Update build.gradle for new Jar File

freddychenyouren2 committed:
Update build.gradle for new Jar File (Master)

freddychenyouren2 committed:
Merge branch 'A-CheckStyle'

# Conflicts:
#	build.gradle

freddychenyouren2 committed:
Add and enable Java assertions

freddychenyouren2 committed:
Improve on Code Quality

There are a few parts in the code that can be improved for readability
and adhering to coding standards

Thus there are changes that have to be made to the code

Here are the changes:
* Format Javadoc descriptions and parameters
* Attempts to break long methods into more but shorter methods
in Storage class
* Removing unnecessary code-level comments
* Removing redundant exception throws in method signatures

freddychenyouren2 committed:
Merge pull request #1 from freddychenyouren2/A-Assertions

Add and enable Java assertions (PR A-Assertion)

freddychenyouren2 committed:
Merge branch 'master' into A-CodingQuality

freddychenyouren2 committed:
Merge pull request #2 from freddychenyouren2/A-CodingQuality

Improve on Code Quality

freddychenyouren2 committed:
Implement ViewSchedule

When given a date, return the list of tasks that are happening on that day.

freddychenyouren2 committed:
Update JavaDoc for TaskList file

freddychenyouren2 committed:
Merge pull request #3 from freddychenyouren2/branch-B-ViewSchedule

Implement ViewSchedule

freddychenyouren2 committed:
Implement greet and farewell methods for GUI

GUI did not show greeting message and did not close after farewell() is called

freddychenyouren2 committed:
Update README.md for User Guide

freddychenyouren2 committed:
Include Ui.png for ip showcase on project website

freddychenyouren2 committed:
Update README.md

freddychenyouren2 committed:
Update README.md

freddychenyouren2 committed:
Update REAMDE.MD for Command Summary

freddychenyouren2 committed:
Reformat Feature subtitles in README.md User Guide

freddychenyouren2 committed:
Add more JUnit tests on TaskList

TaskList class contains some major methods of SeeWhyAre that did not have automated testing

Let's implement test cases in a test file for the TaskList class to test the method functionalities.

freddychenyouren2 committed:
Update Shadow JAR file name into 'seeWhyAre'

freddychenyouren2 committed:
Refactor Parser file to have shorter method length

freddychenyouren2 committed:
Remove dead code and improve Javadoc

Improvements done specified by SoC SE Bot Red

freddychenyouren2 committed:
Refactor Storage class to have shorter methods

loadTasks() was considered too long

Let's refactor the code by introducing multiple helper functions
to handle loading different types of tasks.

freddychenyouren2 committed:
Refactor code for shorter methods and follow Java Checkstyle

freddychenyouren2 committed:
Fix bug on taskListTest file creation

freddychenyouren2 committed:
Fix bug on encoding saved tasks

There were observations on different devices that the file duke.txt
containing saved tasks has a newline character at the beginning of the file.

Another observation seen is that there is no newline character after the first task
is encoded, causing the second task to be
written on the same line as the first task

The loadTask() function subsequently reads the line as a corrupted line.
Some reported issues relaunching the app.

This has been corrected and SeeWhyAre app should be launching normally.

freddychenyouren2 committed:
Reduce user icon image size in chatbot GUI

JAR file was reportedly too big (28.8MB)
the typical JAR file size for iP is 9-11MB

The issue was not in excessive usage of external library
But rather the image of Duke and User have
exceedingly high resolution

Let's reduce the file size of both images to
significantly reduce the JAR file produced

freddychenyouren2 committed:
Fix bug on incorrect task storage handling after delete command
